cleverhans/attacks/bapp.py

```diff
@@ -0,0 +1,573 @@
+                                          decision_function, self.shape,
+                                          self.constraint, self.theta)
+
+    for j in np.arange(self.num_iterations):
+      current_iteration = j + 1
+
+      # Choose delta.
+      delta = select_delta(dist_post_update, current_iteration,
+                           self.clip_max, self.clip_min, self.d,
+                           self.theta, self.constraint)
+
+      # Choose number of evaluations.
+      num_evals = int(min([self.initial_num_evals * np.sqrt(j+1),
+                           self.max_num_evals]))
+
+      # approximate gradient.
+      gradf = approximate_gradient(decision_function, perturbed, num_evals,
+                                   delta, self.constraint, self.shape,
+                                   self.clip_min, self.clip_max)
+      if self.constraint == 'linf':
+        update = np.sign(gradf)
+      else:
+        update = gradf
+
+      # search step size.
+      if self.stepsize_search == 'geometric_progression':
+        # find step size.
+        epsilon = geometric_progression_for_stepsize(perturbed,
+                                                     update, dist, decision_function, current_iteration)
+
+        # Update the sample.
+        perturbed = clip_image(perturbed + epsilon * gradf,
+                               self.clip_min, self.clip_max)
+
+        # Binary search to return to the boundary.
+        perturbed, dist_post_update = binary_search_batch(sample,
+                                                          perturbed[None],
+                                                          decision_function,
+                                                          self.shape,
+                                                          self.constraint,
+                                                          self.theta)
+
+      elif self.stepsize_search == 'grid_search':
+        # Grid search for stepsize.
+        epsilons = np.logspace(-4, 0, num=20, endpoint=True) * dist
+        epsilons_shape = [20] + len(self.shape) * [1]
+        perturbeds = perturbed + epsilons.reshape(epsilons_shape) * update
+        idx_perturbed = decision_function(perturbeds)
+
+        if np.sum(idx_perturbed) > 0:
+          # Select the perturbation that yields the minimum distance # after binary search.
+          perturbed, dist_post_update = binary_search_batch(sample,
+                                                            perturbeds[idx_perturbed], decision_function, self.shape,
+                                                            self.constraint, self.theta)
+
+      # compute new distance.
+      dist = compute_distance(perturbed, sample, self.constraint)
+      if self.verbose:
+        print('iteration: {:d}, {:s} distance {:.4E}'.format(
+            j+1, self.constraint, dist))
+
+    perturbed = np.expand_dims(perturbed, 0)
+    return perturbed
+
+
+def clip_image(image, clip_min, clip_max):
```

Comment:
You can probably reuse: https://github.com/tensorflow/cleverhans/blob/f70ca7e000dadd6ace5aeff15bba0e960e8c1384/cleverhans/utils_tf.py#L540

Reply:
clip_by_value only applies to clipping by 0-dim tensor. clip_image is capable of clipping by a tensor with the same shape as the image. Also, as the wrapper is used, it might be more consistent to not use tf operators in the function.